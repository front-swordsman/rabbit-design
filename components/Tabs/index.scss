@import '../styles/themes/default';

$tabs: 'snake-tabs';

.#{$tabs} {
  box-sizing: border-box;

  &-wrap {
    display: flex;
    position: relative;
    background: #fff;
    overflow-x: scroll;

    &::-webkit-scrollbar {
      display: none; // 移除滚动条
    }

    &-top {
      height: 40px;
      @include one-px-border(bottom, solid, #ddd);
    }

    &-bottom {
      height: 40px;
      @include one-px-border(top, solid, #ddd);
    }

    &-left {
      flex-direction: column;
      @include one-px-border(right, solid, #ddd);
    }

    &-right {
      flex-direction: column;
      @include one-px-border(left, solid, #ddd);
    }

    &-item {
      position: relative;
      display: flex;
      justify-content: center;
      align-items: center;
      color: #666;
      padding: 6px 12px;
      cursor: pointer;

      /* 禁用态 */
      &-disabled {
        color: rgba(0, 0, 0, 0.25);
        cursor: not-allowed;
      }
    }

    &-scroll {
      flex-shrink: 0;
    }

    &-active {
      color: $brand;
    }

    &-underline {
      position: absolute;
      border: 1px solid $brand;
      z-index: 10;

      &-top {
        bottom: 0;
        transition: left 0.3s $easeInOut, width 0.3s $easeInOut;
      }

      &-bottom {
        top: 0;
        transition: left 0.3s $easeInOut, width 0.3s $easeInOut;
      }

      &-left {
        right: 0;
        transition: top 0.3s $easeInOut, width 0.3s $easeInOut;
      }

      &-right {
        left: 0;
        transition: top 0.3s $easeInOut, width 0.3s $easeInOut;
      }
    }
  }

  &-left,
  &-right {
    display: flex;
  }

  &-card {
    &-top {
      background: #fafafa;
      border: 1px solid #e8e8e8;
      border-radius: 4px 4px 0 0;
      border-bottom: 1px solid #fff;
      margin-right: 2px;

      &-current {
        background: #fff;
        border-bottom: 1px solid #fff;
        z-index: 1;
      }
    }
  }
}
