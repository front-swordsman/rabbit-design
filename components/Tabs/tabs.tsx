import * as React from 'react'
import { TabsProps, TabsItemProps } from 'types/tabs.d'
import cx from 'classnames'

const { useEffect, useCallback } = React

const defaultProps = {
  prefixCls: 'snake-tabs',
  tabBarPosition: 'top',
  type: 'line',
  tabBarUnderlineColor: '#01C1B2'
}

function Tabs(userProps: TabsProps, ref: React.RefObject<any>) {
  const props = {
    ...defaultProps,
    ...userProps
  }

  const {
    prefixCls,
    tabBarPosition,
    activeTab,
    tabBarUnderlineColor,
    options,
    onChange,
    type,
    className,
    style
  } = props

  let underline: any = null
  let tabObj: any = {}
  let tabWrap: any = null
  let tabItemObj: any = {}

  useEffect(() => {
    handleUnderline()
  }, [activeTab])

  // 处理下划线样式
  const handleUnderline = () => {
    if (tabBarPosition === 'top' || tabBarPosition === 'bottom') {
      underline && (underline.style.width = `${getUnderlineWidth(activeTab)}px`)
      underline && (underline.style.left = `${getLeft(activeTab)}px`)
    } else if (tabBarPosition === 'left' || tabBarPosition === 'right') {
      underline && (underline.style.height = `${getUnderlineHeight(activeTab)}px`)
      underline && (underline.style.top = `${getTop(activeTab)}px`)
    }
  }

  // 获取下划线宽度
  const getUnderlineWidth = useCallback(
    (current: number) =>
      tabObj[`tab${current}`] && tabObj[`tab${current}`].getBoundingClientRect().width,
    [activeTab]
  )

  // 获取下划线高度
  const getUnderlineHeight = useCallback(
    (current: number) =>
      tabItemObj[`tab${current}`] && tabItemObj[`tab${current}`].getBoundingClientRect().height,
    [activeTab]
  )

  // 获取当前激活 tab 距离左侧的距离
  const getLeft = useCallback(
    (current: number) => {
      if (tabObj[`tab${current}`] && tabWrap) {
        return (
          tabObj[`tab${current}`].getBoundingClientRect().left -
          tabWrap.getBoundingClientRect().left
        )
      } else {
        return 0
      }
    },
    [activeTab]
  )

  // 获取当前激活 tab 距离上侧的距离
  const getTop = useCallback(
    (current: number) => {
      if (tabObj[`tab${current}`] && tabWrap) {
        return (
          tabItemObj[`tab${current}`].getBoundingClientRect().top -
          tabWrap.getBoundingClientRect().top
        )
      } else {
        return 0
      }
    },
    [activeTab]
  )

  // 更改 Tab
  const changeTab = (e: React.MouseEvent<HTMLElement>, index: number, r: TabsItemProps) => {
    // 禁用 tab 点击
    if (r.disabled) return

    if (index !== activeTab) {
      onChange && onChange(index, e)
    }
  }

  const renderTab = () => {
    return (
      <div
        className={cx(`${prefixCls}-wrap`, `${prefixCls}-wrap-${tabBarPosition}`)}
        ref={(r: any) => (tabWrap = r)}
      >
        {options.map((r: TabsItemProps, index: number) => {
          return (
            <div
              className={cx(`${prefixCls}-wrap-item`, {
                [`${prefixCls}-wrap-item-disabled`]: r.disabled === true,
                [`${prefixCls}-card-${tabBarPosition}`]: type === 'card',
                [`${prefixCls}-card-${tabBarPosition}-current`]:
                  type === 'card' && index === activeTab
              })}
              onClick={e => changeTab(e, index, r)}
              key={index}
              ref={ele => {
                tabItemObj[`tab${index}`] = ele
              }}
            >
              <div
                className={cx({
                  [`${prefixCls}-wrap-active`]: activeTab === index
                })}
                ref={ele => {
                  tabObj[`tab${index}`] = ele
                }}
              >
                {r.title || ''}
              </div>
            </div>
          )
        })}
        {renderUnderline()}
      </div>
    )
  }

  const renderUnderline = () => {
    if (type === 'line') {
      return (
        <div
          className={cx(
            `${prefixCls}-wrap-underline`,
            `${prefixCls}-wrap-underline-${tabBarPosition}`
          )}
          style={{ border: `1px solid ${tabBarUnderlineColor}` }}
          ref={ele => {
            underline = ele
          }}
        />
      )
    } else {
      return null
    }
  }

  const renderContent = () => {
    if (activeTab >= 0 && options[activeTab].content) {
      return options[activeTab].content
    } else {
      return null
    }
  }

  function layout() {
    switch (tabBarPosition) {
      case 'left':
        return (
          <div className={`${prefixCls}-left`}>
            {renderTab()}
            {renderContent()}
          </div>
        )
      case 'right':
        return (
          <div className={`${prefixCls}-right`}>
            {renderContent()}
            {renderTab()}
          </div>
        )
      case 'bottom':
        return (
          <div>
            {renderContent()}
            {renderTab()}
          </div>
        )
      default:
        return (
          <div>
            {renderTab()}
            {renderContent()}
          </div>
        )
    }
  }

  function getClassName() {
    return cx(`${prefixCls}`, className)
  }

  return (
    <div className={getClassName()} style={style} ref={ref}>
      {layout()}
    </div>
  )
}

export default React.forwardRef(Tabs)
