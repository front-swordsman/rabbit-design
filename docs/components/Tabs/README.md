---
imports: import TabsDemoCode from '!raw-loader!./simpleTabs.tsx';
  import TabDemo from './simpleTabs.tsx';
---

# Tabs 标签页

选项卡切换组件。

## 简单示例

:::demo
<Block des="Tabs 示例" code={TabsDemoCode}>
<TabDemo />
</Block>
:::

## Api

| 属性                 | 说明                           | 类型                              | 默认值   | 必填 |
| :------------------- | :----------------------------- | :-------------------------------- | :------- | :--- |
| options              | tabs 中的展示项                | TabsItemProps[]                   | []       | 否   |
| activeTab            | tab 当前选中页                 | number                            | 0        | 否   |
| onChange             | Tab 切换时触发                 | Function                          | --       | 否   |
| onTabClick           | 点击 Tab 时触发                | Function                          | --       | 否   |
| tabBarPosition       | TabBar 位置                    | 'top' \| 'bottom \| left \| right | top      | 否   |
| tabItemStyle         | tabItem 的样式                 | React.CSSProperties               | --       | 否   |
| type                 | 提供 'line' \| 'card' 两种模式 | string                            | 'line'   | 否   |
| tabBarUnderlineColor | 下划线颜色                     | string                            | '01C1B2' | 否   |

## TabsItemProps Api

| 属性     | 说明                  | 类型                          | 默认值 | 必填 |
| :------- | :-------------------- | :---------------------------- | :----- | :--- |
| title    | Tabs 中的单项         | String \| Number \| ReactNode | --     | 是   |
| content  | Tabs 中的内容         | ReactNode                     | --     | 是   |
| disabled | Tabs 中的单项是否禁用 | boolean                       | false  | 否   |
