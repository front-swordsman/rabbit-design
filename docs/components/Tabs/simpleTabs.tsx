import * as React from 'react'
import Tabs from 'components/Tabs/tabs'
import './demo.scss'

const tabs0 = [
  {
    title: '标签一',
    content: (
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          height: '120px',
          backgroundColor: '#fff'
        }}
      >
        标签一的内容
      </div>
    )
  },
  {
    title: '标签二',
    content: (
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          height: '120px',
          backgroundColor: '#fff'
        }}
      >
        标签二的内容
      </div>
    )
  },
  {
    title: '标签三',
    content: (
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          height: '120px',
          backgroundColor: '#fff'
        }}
      >
        标签三的内容
      </div>
    )
  },
  {
    title: '标签四',
    content: (
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          height: '120px',
          backgroundColor: '#fff'
        }}
      >
        标签四的内容
      </div>
    )
  }
]

const tabs1 = [
  {
    title: '标签一',
    content: (
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          height: '120px',
          backgroundColor: '#fff'
        }}
      >
        标签一的内容
      </div>
    )
  },
  {
    title: '标签二',
    content: (
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          height: '120px',
          backgroundColor: '#fff'
        }}
      >
        标签二的内容
      </div>
    ),
    disabled: true
  },
  {
    title: '标签三',
    content: (
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          height: '120px',
          backgroundColor: '#fff'
        }}
      >
        标签三的内容
      </div>
    )
  },
  {
    title: '标签四',
    content: (
      <div
        style={{
          display: 'flex',
          alignItems: 'center',
          justifyContent: 'center',
          height: '120px',
          backgroundColor: '#fff'
        }}
      >
        标签四的内容
      </div>
    )
  }
]

export default class SimpleTab extends React.Component {
  state = {
    page0: 0,
    page1: 0,
    page2: 0,
    page3: 0
  }

  onChange = (index: number, _e: any, order: any) => {
    this.setState({
      [`page${order}`]: index
    })
  }

  render() {
    const { page0, page1, page2, page3 } = this.state

    return (
      <>
        <div>线性模式</div>
        <Tabs
          options={tabs0}
          activeTab={page0}
          onChange={(index: number, e: any) => this.onChange(index, e, 0)}
          style={{ width: 300 }}
        />
        <div>垂直模式</div>
        <Tabs
          options={tabs0}
          activeTab={page1}
          onChange={(index: number, e: any) => this.onChange(index, e, 1)}
          tabBarPosition="left"
        />
        <div>卡片模式(暂不提供卡片模式下的垂直模式)</div>
        <Tabs
          options={tabs0}
          activeTab={page2}
          onChange={(index: number, e: any) => this.onChange(index, e, 2)}
          type="card"
          style={{ width: 300 }}
        />
        <div>禁用</div>
        <Tabs
          options={tabs1}
          activeTab={page3}
          onChange={(index: number, e: any) => this.onChange(index, e, 3)}
          style={{ width: 300 }}
        />
      </>
    )
  }
}
