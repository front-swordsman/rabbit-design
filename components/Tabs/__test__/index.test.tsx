import * as React from 'react'
import { render, fireEvent } from 'react-testing-library'
import Tabs from '../index'

const { useState } = React

const tabs = [
  {
    title: '标签一',
    content: <div>标签一的内容</div>
  },
  {
    title: '标签二',
    content: <div>标签二的内容</div>
  },
  {
    title: '标签三',
    content: <div>标签三的内容</div>,
    disabled: true
  }
]

describe('Tabs Test', () => {
  beforeEach(() => {
    document.body.innerHTML = ''
  })
  it('snapshot', () => {
    const tabsRenderTop = render(<Tabs options={tabs} />)
    const tabsRenderBottom = render(<Tabs options={tabs} tabBarPosition="bottom" />)
    const tabsRenderLeft = render(<Tabs options={tabs} tabBarPosition="left" />)
    const tabsRenderRight = render(<Tabs options={tabs} tabBarPosition="right" />)
    expect(tabsRenderTop.container.innerHTML).toMatchSnapshot()
    expect(tabsRenderBottom.container.innerHTML).toMatchSnapshot()
    expect(tabsRenderLeft.container.innerHTML).toMatchSnapshot()
    expect(tabsRenderRight.container.innerHTML).toMatchSnapshot()
  })
  it('dispatch change tab', async () => {
    const TabDemo = function() {
      const [activeTab, setActiveTab] = useState(0)
      return (
        <Tabs
          options={tabs}
          activeTab={activeTab}
          onChange={(index: number) => setActiveTab(index)}
        />
      )
    }

    const { getByText } = render(<TabDemo />)
    expect(getByText('标签一的内容')).toBeTruthy()
    fireEvent.click(getByText('标签二'))
    expect(getByText('标签二的内容')).toBeTruthy()
    fireEvent.click(getByText('标签三'))
    expect(getByText('标签二的内容')).toBeTruthy()
  })
})
