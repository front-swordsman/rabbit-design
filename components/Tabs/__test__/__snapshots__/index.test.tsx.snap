// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Test snapshot 1`] = `"<div class=\\"snake-tabs\\"><div><div class=\\"snake-tabs-wrap snake-tabs-wrap-top\\"><div class=\\"snake-tabs-wrap-item\\"><div class=\\"\\">标签一</div></div><div class=\\"snake-tabs-wrap-item\\"><div class=\\"\\">标签二</div></div><div class=\\"snake-tabs-wrap-item snake-tabs-wrap-item-disabled\\"><div class=\\"\\">标签三</div></div><div class=\\"snake-tabs-wrap-underline snake-tabs-wrap-underline-top\\" style=\\"border: 1px solid #01c1b2; left: 0px;\\"></div></div></div></div>"`;

exports[`Tabs Test snapshot 2`] = `"<div class=\\"snake-tabs\\"><div><div class=\\"snake-tabs-wrap snake-tabs-wrap-bottom\\"><div class=\\"snake-tabs-wrap-item\\"><div class=\\"\\">标签一</div></div><div class=\\"snake-tabs-wrap-item\\"><div class=\\"\\">标签二</div></div><div class=\\"snake-tabs-wrap-item snake-tabs-wrap-item-disabled\\"><div class=\\"\\">标签三</div></div><div class=\\"snake-tabs-wrap-underline snake-tabs-wrap-underline-bottom\\" style=\\"border: 1px solid #01c1b2; left: 0px;\\"></div></div></div></div>"`;

exports[`Tabs Test snapshot 3`] = `"<div class=\\"snake-tabs\\"><div class=\\"snake-tabs-left\\"><div class=\\"snake-tabs-wrap snake-tabs-wrap-left\\"><div class=\\"snake-tabs-wrap-item\\"><div class=\\"\\">标签一</div></div><div class=\\"snake-tabs-wrap-item\\"><div class=\\"\\">标签二</div></div><div class=\\"snake-tabs-wrap-item snake-tabs-wrap-item-disabled\\"><div class=\\"\\">标签三</div></div><div class=\\"snake-tabs-wrap-underline snake-tabs-wrap-underline-left\\" style=\\"border: 1px solid #01c1b2; top: 0px;\\"></div></div></div></div>"`;

exports[`Tabs Test snapshot 4`] = `"<div class=\\"snake-tabs\\"><div class=\\"snake-tabs-right\\"><div class=\\"snake-tabs-wrap snake-tabs-wrap-right\\"><div class=\\"snake-tabs-wrap-item\\"><div class=\\"\\">标签一</div></div><div class=\\"snake-tabs-wrap-item\\"><div class=\\"\\">标签二</div></div><div class=\\"snake-tabs-wrap-item snake-tabs-wrap-item-disabled\\"><div class=\\"\\">标签三</div></div><div class=\\"snake-tabs-wrap-underline snake-tabs-wrap-underline-right\\" style=\\"border: 1px solid #01c1b2; top: 0px;\\"></div></div></div></div>"`;
